packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts

```diff
@@ -0,0 +1,305 @@
+    this.roleName = this.role.roleName;
+    this.stack = this.role.stack;
+
+    Dependable.implement(this, {
+      dependencyRoots: [this.role],
+    });
+
+    // add a single PolicySynthesizer under the `App` scope
+    this.policySynthesizer = (this.node.root.node.tryFindChild(POLICY_SYNTHESIZER_ID)
+      ?? new PolicySynthesizer(this.node.root)) as PolicySynthesizer;
+    this.policySynthesizer.addRole(this.node.path, {
+      roleName: props.missing ? 'missing role' : this.roleName,
+      managedPolicies: this.managedPolicies,
+      policyStatements: this.policyStatements,
+      assumeRolePolicy: Stack.of(this).resolve(props.assumeRolePolicy?.toJSON()?.Statement),
+    });
+  }
+
+  public attachInlinePolicy(policy: Policy): void {
+    const statements = policy.document.toJSON()?.Statement;
+    if (statements && Array.isArray(statements)) {
+      statements.forEach(statement => {
+        this.policyStatements.push(this.resolveJsonObject(statement));
+      });
+    }
+  }
+
+  public addManagedPolicy(policy: IManagedPolicy): void {
+    this.managedPolicies.push(this.resolveReferences(policy.managedPolicyArn));
+  }
+
+  public addToPolicy(statement: PolicyStatement): boolean {
+    this.policyStatements.push(this.resolveReferences(statement.toStatementJson()));
+    return false;
+  }
+
+  public addToPrincipalPolicy(statement: PolicyStatement): AddToPrincipalPolicyResult {
+    this.addToPolicy(statement);
+    // If we return `false`, the grants will try to add the statement to the resource
+    // (if possible).
+    return { statementAdded: true, policyDependable: new DependencyGroup() };
+  }
+
+  public grant(grantee: IPrincipal, ...actions: string[]): Grant {
+    return this.role.grant(grantee, ...actions);
+  }
+
+  public grantPassRole(grantee: IPrincipal): Grant {
+    return this.role.grantPassRole(grantee);
+  }
+
+  public grantAssumeRole(identity: IPrincipal): Grant {
+    return this.role.grantAssumeRole(identity);
+  }
+
+  /**
+   * Resolve any references and replace with a more user friendly value. This is the value
+   * that will appear in the report, so instead of getting something like this (not very useful):
+   *
+   *     "Resource": {
+   *       "Fn::Join": [
+   *         "",
+   *         [
+   *           "arn:",
+   *           {
+   *             "Ref": "AWS::Partition"
+   *           },
+   *           ":iam::",
+   *           {
+   *             "Ref": "AWS::AccountId"
+   *           },
+   *           ":role/Role"
+   *         ]
+   *       ]
+   *     }
+   *
+   * We will instead get:
+   *
+   *     "Resource": "arn:(PARTITION):iam::(ACCOUNT):role/Role"
+   *
+   * Or if referencing a resource attribute
+   *
+   *     "Resource": {
+   *       "Fn::GetAtt": [
+   *         "SomeResource",
+   *         "Arn"
+   *       ]
+   *     }
+   *
+   * Becomes
+   *
+   *     "(Path/To/SomeResource.Arn)"
+   */
+  private resolveReferences(ref: any): any {
```

Comment:
Shouldn't this be the policysynthesizer's job? Feels like part of rendering.

Reply:
yeah i'll move it. For some reason I forgot I can use `Tokenization.resolve` instead of `Stack.resolve`  ha